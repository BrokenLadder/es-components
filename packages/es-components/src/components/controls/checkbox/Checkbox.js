import React from 'react';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components';

import CheckboxLabel from '../../containers/checkboxLabel/CheckboxLabel';

const CheckboxInput = styled.input`
  clip: rect(0, 0, 0, 0);
  pointer-events: none;
  position: absolute;

  &:focus ~ .es-checkbox__fill {
    box-shadow: 0 0 3px 3px ${props => props.theme.colors.inputFocus};
    &:after {
      border-color: ${({ checked, theme }) =>
        checked ? theme.colors.white : theme.colors.gray3};
    }
  }
`;

const CheckboxDisplay = styled.span`
  background: ${props => props.theme.colors.white};
  border: 3px solid ${props => props.theme.colors.gray8};
  border-radius: 4px;
  box-sizing: border-box;
  cursor: pointer;
  height: 25px;
  left: 10px;
  position: absolute;
  top: 0.55em;
  transition: all 0.25s linear;
  width: 25px;

  @media (min-width: ${props => props.theme.screenSize.tablet}) {
    left: 0;
    top: 0.35em;
  }

  &:after {
    background: transparent;
    border: 3px solid ${props => props.theme.colors.white};
    border-right: none;
    border-top: none;
    box-sizing: border-box;
    content: '';
    display: block;
    height: 9px;
    margin: 3px 0 0 2px;
    transform: rotate(-45deg);
    transition: border 0.25s linear;
    width: 15px;
  }
`;

const AdditionalHelpContent = styled.div`
  font-size: ${props => props.theme.sizes.baseFontSize};
  font-weight: 400;
  margin: 10px 0 10px 0;
  position: relative;
  right: 32px;
`;

function Checkbox({
  name,
  labelText,
  validationState,
  additionalHelpContent,
  theme,
  ...checkboxProps
}) {
  const helpId = additionalHelpContent ? `${name}-help` : undefined;
  const additionalHelp = additionalHelpContent && (
    <AdditionalHelpContent
      id={helpId}
      className="es-checkbox__help"
      validationState={validationState}
    >
      {additionalHelpContent}
    </AdditionalHelpContent>
  );

  return (
    <CheckboxLabel
      className="es-checkbox"
      validationState={validationState}
      checked={checkboxProps.checked}
      disabled={checkboxProps.disabled}
    >
      <CheckboxInput
        name={name}
        type="checkbox"
        focusBorderColor={theme.colors.inputFocus}
        {...checkboxProps}
      />
      <CheckboxDisplay className="es-checkbox__fill" />
      {labelText}
      {additionalHelp}
    </CheckboxLabel>
  );
}

Checkbox.propTypes = {
  /** The name of the checkbox */
  name: PropTypes.string.isRequired,
  labelText: PropTypes.string.isRequired,
  /** Display checkbox with contextual state colorings */
  validationState: PropTypes.oneOf(['default', 'success', 'warning', 'danger']),
  /** Content to display underneath the check box */
  additionalHelpContent: PropTypes.node,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object.isRequired
};

Checkbox.defaultProps = {
  validationState: 'default',
  additionalHelpContent: undefined
};

export default withTheme(Checkbox);
