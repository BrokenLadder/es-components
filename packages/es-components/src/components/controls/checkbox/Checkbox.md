```
<div>
  <Checkbox name="uncheckedByDefault" labelText="I am unchecked by default" />
  <Checkbox name="checkedByDefault" labelText="I am checked by default" checked />
  <Checkbox name="disabled" labelText="I am disabled" disabled />
  <Checkbox name="uncheckedByDefault" labelText="I am checked and disabled" checked disabled />
</div>
```

Passing a function will execute with the current selection status of the Checkbox.

```
class CheckboxExample extends React.Component {
  constructor() {
    this.state = { 
      isAppleChecked: false,
      isBananaChecked: false,
      isStrawberryChecked: true
    }

    this.toggleCheckbox = this.toggleCheckbox.bind(this);
  }

  toggleCheckbox(key) {
    return event => {
      this.setState({ [key]: event.target.checked });
    }
  }

  render() {
    return (
      <div>
        <Checkbox 
          name="Apple"
          labelText="Apple"
          value="Apple"
          checked={this.state.isAppleChecked}
          onClick={this.toggleCheckbox('isAppleChecked')}
        />
        <Checkbox 
          name="Banana"
          labelText="Banana"
          value="Banana"
          checked={this.state.isBananaChecked}
          onClick={this.toggleCheckbox('isBananaChecked')}
        />
        <Checkbox 
          name="Strawberry"
          labelText="Strawberry"
          value="Strawberry"
          checked={this.state.isStrawberryChecked}
          onClick={this.toggleCheckbox('isStrawberryChecked')}
          disabled
        />
      </div>
    );
  }
}

<CheckboxExample />
```
### Validation States
```
class CheckboxExample extends React.Component {
  constructor() {
    this.state = { 
      isAppleChecked: false,
      isBananaChecked: false,
      isStrawberryChecked: false
    }

    this.toggleCheckbox = this.toggleCheckbox.bind(this);
  }

  toggleCheckbox(key) {
    return event => {
      this.setState({ [key]: event.target.checked });
    }
  }

  render() {
    return (
      <div>
        <Checkbox 
          name="Apple_Success"
          labelText="Apple"
          value="Apple"
          checked={this.state.isAppleChecked}
          onClick={this.toggleCheckbox('isAppleChecked')}
          validationState="success"
          additionalHelpContent="When validationState is set to success."
        />
        <Checkbox 
          name="Banana_Success"
          labelText="Banana"
          value="Banana"
          checked={this.state.isBananaChecked}
          onClick={this.toggleCheckbox('isBananaChecked')}
          validationState="warning"
          additionalHelpContent="When validationState is set to warning."
        />
        <Checkbox 
          name="Strawberry_danger"
          labelText="Strawberry"
          value="Strawberry"
          checked={this.state.isStrawberryChecked}
          onClick={this.toggleCheckbox('isStrawberryChecked')}
          validationState="danger"
          additionalHelpContent="When validationState is set to danger."
        />
      </div>
    );
  }
}

<CheckboxExample />
```
