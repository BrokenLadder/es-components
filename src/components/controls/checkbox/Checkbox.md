```
<div>
  <Checkbox labelText="I am unchecked by default" />
  <Checkbox labelText="I am checked by default" initiallyChecked />
  <Checkbox labelText="I am disabled" disabled />
  <Checkbox labelText="I am checked and disabled" initiallyChecked disabled />
</div>
```

Passing a function will execute with the current selection status of the Checkbox.

```
function printIsChecked(isChecked) {
  console.log(`Is checked? ${isChecked}`);
}

<Checkbox labelText="Check the console when you check and uncheck me" checkedValueChanged={printIsChecked} />
```