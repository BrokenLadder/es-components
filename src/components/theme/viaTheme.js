const white = '#fff';
const gray1 = '#f4f4f4';
const gray2 = '#ecedee';
const gray3 = '#e4e4e4';
const gray4 = '#d8d8d8';
const gray5 = '#c5c5c5';
const gray6 = '#979797';
// const gray7 = '#707070';
const gray8 = '#585858';
const gray9 = '#444';
const black = '#000';

const grayLightest = gray1;
const grayLighter = gray2;
const grayLight = gray3;
const gray = gray5;
const grayDark = gray6;
const grayDarker = gray8;
const grayDarkest = gray9;

const dflt = '#d8d8d8';
const dfltHover = '#bfbfbf';
const primary = '#007fa7';
const primaryHover = '#005874';
const info = '#069';
const infoHover = '#004466';
const success = '#298544';
const successHover = '#1d5e30';
const warning = '#b35f00';
const warningHover = '#804400';
const danger = '#c00';
const dangerHover = '#990000';
const advisor = warning;

const boxShadowLight = 'rgba(0, 0, 0, 0.075)';
const boxShadowDark = 'rgba(0, 0, 0, 0.5)';

const inputBorder = gray4;
const inputFocus = '#83bffc';
const inputBoxShadow = 'rgba(102, 175, 233, 0.6)';

// Datepicker colors
const navArrow = gray;
const navArrowHover = grayDark;
const selected = '#216ba5';
const hover = '#1d5d90';
const keyboard = '#2a87d0';
const inRange = 'rgba(33, 107, 165, 0.5)';
const highlight = '#3dcc4a';
const highlightHover = '#32be3f';

const theme = {
  colors: {
    black,
    grayDarkest,
    grayDarker,
    grayDark,
    gray,
    grayLight,
    grayLighter,
    grayLightest,
    white,
    dflt,
    dfltHover,
    primary,
    primaryHover,
    success,
    successHover,
    info,
    infoHover,
    warning,
    warningHover,
    danger,
    dangerHover,
    advisor,
    boxShadowLight,
    boxShadowDark,
    inputFocus
  },
  datepickerColors: {
    navArrow,
    navArrowHover,
    selected,
    hover,
    keyboard,
    inRange,
    highlight,
    highlightHover
  },
  buttonStyles: {
    buttonsNormal: {
      primary: {
        bgColor: primary,
        textColor: white,
        hoverBgColor: primaryHover,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        boxShadowColor: primaryHover
      },
      secondary: {
        bgColor: dflt,
        textColor: white,
        hoverBgColor: dfltHover,
        hoverTextColor: white,
        activeBgColor: dfltHover,
        activeTextColor: white,
        boxShadowColor: dfltHover
      },
      success: {
        bgColor: success,
        textColor: white,
        hoverBgColor: successHover,
        hoverTextColor: white,
        activeBgColor: successHover,
        activeTextColor: white,
        boxShadowColor: successHover
      },
      info: {
        bgColor: info,
        textColor: white,
        hoverBgColor: infoHover,
        hoverTextColor: white,
        activeBgColor: infoHover,
        activeTextColor: white,
        boxShadowColor: infoHover
      },
      warning: {
        bgColor: warning,
        textColor: white,
        hoverBgColor: warningHover,
        hoverTextColor: white,
        activeBgColor: warningHover,
        activeTextColor: white,
        boxShadowColor: warningHover
      },
      danger: {
        bgColor: danger,
        textColor: white,
        hoverBgColor: dangerHover,
        hoverTextColor: white,
        activeBgColor: dangerHover,
        activeTextColor: white,
        boxShadowColor: dangerHover
      }
    },
    buttonsAlt: {
      primary: {
        bgColor: white,
        textColor: primary,
        hoverBgColor: primary,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        borderColor: primary
      },
      gray: {
        bgColor: white,
        textColor: '#63666a',
        hoverBgColor: '#63666a',
        hoverTextColor: white,
        activeBgColor: '#4a4d50',
        activeTextColor: white,
        borderColor: '#63666a'
      },
      magenta: {
        bgColor: white,
        textColor: '#c110a0',
        hoverBgColor: '#c110a0',
        hoverTextColor: white,
        activeBgColor: '#920C79',
        activeTextColor: white,
        borderColor: '#c110a0'
      },
      violet: {
        bgColor: white,
        textColor: '#702082',
        hoverBgColor: '#702082',
        hoverTextColor: white,
        activeBgColor: '#4d1659',
        activeTextColor: white,
        borderColor: '#702082'
      }
    }
  },
  // buttonSizes should always have default, lg, sm, xs
  buttonSizes: {
    lg: {
      borderRadius: '3px',
      fontSize: '26px',
      padding: '8px 20px 5px'
    },
    default: {
      borderRadius: '2px',
      fontSize: '18px',
      padding: '5px 15px 4px'
    },
    sm: {
      borderRadius: '2px',
      fontSize: '14px',
      padding: '4px 10px 3px'
    },
    xs: {
      borderRadius: '2px',
      fontSize: '12px',
      padding: '3px 10px 2px'
    }
  },
  validationIconName: {
    success: 'ok',
    warning: 'warning-sign',
    danger: 'remove'
  },
  validationTextColor: {
    success,
    warning: warningHover,
    danger
  },
  validationInputColor: {
    success: {
      borderColor: success,
      boxShadow: `inset 0 0 5px ${success}`,
      focusBorderColor: success,
      focusBoxShadow: `0 0 6px ${success}`
    },
    warning: {
      borderColor: warning,
      boxShadow: `inset 0 0 5px ${warning}`,
      focusBorderColor: warning,
      focusBoxShadow: `0 0 6px ${warning}`
    },
    danger: {
      borderColor: danger,
      boxShadow: `inset 0 0 5px ${danger}`,
      focusBorderColor: danger,
      focusBoxShadow: `0 0 6px ${danger}`
    },
    default: {
      borderColor: inputBorder,
      boxShadow: `inset 0 1px 1px ${boxShadowLight}`,
      focusBorderColor: inputFocus,
      focusBoxShadow: `0 0 8px ${inputBoxShadow}`
    }
  },
  sizes: {
    baseFontSize: 18,
    baseLineHeight: 24
  },
  screenSize: {
    retina: '320px',
    phone: '480px',
    tablet: '768px',
    desktop: '992px',
    widescreen: '1200px'
  }
};

export default theme;
