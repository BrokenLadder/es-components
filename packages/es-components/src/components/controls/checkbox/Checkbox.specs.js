/* eslint-env jest */

import React from 'react';
import { render, cleanup } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';

import Checkbox from './Checkbox';

beforeEach(cleanup);

it('executes onClick prop when label is clicked', () => {
  const onClick = jest.fn();

  const { getByLabelText } = render(
    <ThemeProvider theme={viaTheme}>
      <Checkbox name="test" labelText="test label" onClick={onClick} />
    </ThemeProvider>
  );

  getByLabelText('test label').click();
  expect(onClick).toHaveBeenCalled();
});

it('cannot be clicked when disabled', () => {
  const onClick = jest.fn();
  const { container, getByLabelText } = render(
    <ThemeProvider theme={viaTheme}>
      <Checkbox name="test" labelText="test label" onClick={onClick} disabled />
    </ThemeProvider>
  );

  getByLabelText('test label').click();
  expect(onClick).not.toHaveBeenCalled();
  expect(container.querySelector('input')).toBeDisabled();
});

it('will render additional help when passed', () => {
  const { queryByText } = render(
    <ThemeProvider theme={viaTheme}>
      <Checkbox
        name="test"
        labelText="test label"
        onClick={jest.fn()}
        additionalHelpContent="help me"
      />
    </ThemeProvider>
  );

  expect(queryByText('help me')).not.toBeNull();
});

it('will not render additional help when not passed', () => {
  const { queryByText } = render(
    <ThemeProvider theme={viaTheme}>
      <Checkbox name="test" labelText="test label" onClick={jest.fn()} />
    </ThemeProvider>
  );

  expect(queryByText('help me')).toBeNull();
});
