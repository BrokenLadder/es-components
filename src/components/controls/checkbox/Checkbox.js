import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { noop } from 'lodash';

import Label from '../Label';
import { sizes, colors, inputColors } from '../../theme';

const CheckboxLabel = Label.extend`
  font-size: ${sizes.baseFontSize}px;
  font-weight: normal;
  text-transform: none;

  > .checkbox-fill {
    background-color: ${props => (props.isChecked ? colors.accent : colors.white)};
    border-color: ${props => (props.isChecked ? colors.accent : colors.grayDark)};

    &:after {
      border-color: ${props => (props.isChecked > colors.white: colors.grayDark)};
    }
  }

  &:hover > .checkbox-fill:after {
    border-color: ${props => (props.isChecked ? colors.white : colors.grayLight)};
  }

  &[disabled] > .checkbox-fill {
    background-color: ${props => (props.isChecked ? colors.gray : colors.white)};
    border-color: ${colors.gray};
    cursor: not-allowed;
    outline: 0;

    &:after {
      border-color: ${colors.white};
    }
  }
`;

const CheckboxInput = styled.input`
  clip: rect(0, 0, 0, 0);
  position: absolute;

  &:focus ~ .checkbox-fill {
    outline: 5px auto ${inputColors.inputDefaultFocus};
  }
`;

const CheckboxWrapper = styled.span`
  background: ${colors.white};
  border: 3px solid ${colors.grayDarker};
  border-radius: 4px;
  box-sizing: border-box;
  cursor: pointer;
  height: 25px;
  position: absolute;
  transition: all .25s linear;
  width: 25px;

  &:after {
    background: transparent;
    border: 3px solid ${colors.white};
    border-right: none;
    border-top: none;
    box-sizing: border-box;
    content: '';
    height: 9px;
    top: 3px;
    left: 2px;
    position: absolute;
    transform: rotate(-45deg);
    transition: border .25s linear;
    width: 15px;
  }
`;

const CheckboxText = styled.span`
  line-height: 1.4;
  margin-left: 30px;
`;

export default class Checkbox extends React.Component {
  static propTypes = {
    labelText: PropTypes.string.isRequired,
    initiallyChecked: PropTypes.bool,
    checkedValueChanged: PropTypes.func,
    disabled: PropTypes.bool
  };

  static defaultProps = {
    initiallyChecked: false,
    disabled: false,
    checkedValueChanged: noop
  };

  constructor(props) {
    super(props);

    this.handleOnClick = this.handleOnClick.bind(this);
  }

  state = {
    isChecked: this.props.initiallyChecked
  };

  componentWillReceiveProps({ initiallyChecked }) {
    this.setState(() => ({ isChecked: initiallyChecked }));
  }

  handleOnClick({ target }) {
    if (!this.props.disabled) {
      this.setState(previousState => ({ isChecked: !previousState.isChecked }));
      this.props.checkedValueChanged(target.checked);
    }
  }

  render() {
    const { labelText, disabled } = this.props;

    return (
      <CheckboxLabel disabled={disabled} isChecked={this.state.isChecked}>
        <CheckboxInput
          type="checkbox"
          checked={this.state.isChecked}
          onChange={this.handleOnClick}
          focusBorderColor={inputColors.inputDefaultFocus}
        />
        <CheckboxWrapper className="checkbox-fill" />
        <CheckboxText>{labelText}</CheckboxText>
      </CheckboxLabel>
    );
  }
}
