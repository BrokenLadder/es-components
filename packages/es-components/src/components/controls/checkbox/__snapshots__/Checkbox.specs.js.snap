// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders as expected 1`] = `
<label
  className="es-checkbox sc-bwzfXH blPwYw"
  disabled={true}
>
  <input
    aria-label={null}
    checked={true}
    className="sc-htpNat llDmUy"
    disabled={true}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
    value=""
  />
  <span
    className="es-checkbox__fill sc-bxivhb kynQKu"
  />
  <span
    aria-hidden={false}
    className="es-checkbox__text sc-ifAKCX kWaOTi"
  >
    Render test
  </span>
</label>
`;
